Show support options from /help/contact by the chat and ticket eligibility accordingly

southp opened the pull request:
## Summary
As we are separating eligibility from the chat support and the ticket support, this PR rearranges the conditional flows that leads to the final support option in `/help/contact`.

This PR is based on #9499 and needs the new endpoint, `/help/tickets/kayako/mine`, to be available.

## Test Plan
### Test for the support eligibilities
1. Logging in as a user with plan subscriptions, `/help/contact` should show you chat support as long as there are operators available.
1. Logging in as a user with plan subscriptions, `/help/contact` should show you email support if there is no operator available.
1. Logging in as a user with a la carte upgrades, `/help/contact` should show you email support no matter there are operators available or not.
1. Logging in as a user with access to business sites but doesn't have any upgrades by itself, `/help/contact` should show you chat support as long as there are operators available.
1. Logging in as a user with only free plan, `/help/contact` should show you forum support.

### Test for the error notice
Unfortunately there doesn't seem to be an easy way to fail the request other than ruining it by code. e.g. Change the `Undocumented.prototype.getKayakoConfiguration` to access an endpoint that doesn't exist.

1. Logging in as a chat support eligible user, and you shouldn't see the error notice even if there is an error.
1. Logging in as a email support eligible user, and you should see the error notice like below:
![2016-12-05 2 55 00](https://cloud.githubusercontent.com/assets/1842898/20876141/d660db72-bafa-11e6-8823-80b47914b764.png)
* image updated after f86599d3d5cdba9d7991c771770deab77f7a4475

southp commented:
Updated with the error screen. cc @evilluendas to see if it is okay.

southp commented:
Now that the new endpoint is deployed, it is ready for review. cc @jordwest @omarjackman

jordwest commented:
About to look through the code, but just wanted to leave a thought first. Looking at this, I wonder if we should default to allowing email support when there's an error like this?

In theory, errors should be rare, and it would be frustrating for paid users to be forced to post the forums, especially if the error is on our end.

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -418,12 +421,18 @@ const HelpContact = React.createClass( {
 		const showHappychatVariation = this.shouldUseHappychat();
 		const showChatVariation = olark.isUserEligible && olark.isOperatorAvailable;
-		const showKayakoVariation = ! showChatVariation && ( olark.details.isConversing || olark.isUserEligible );
+		const showKayakoVariation = ! showChatVariation && ( olark.details.isConversing || isTicketSupportEligible );
```
Would it be worth extracting `showKayakoVariation` into a selector?

Maybe later even something like:

```js
function getEligibleSupportLevel( state ) {
  if ( ... ) {
    return HELP_LEVEL_CHAT;
  } else if ( ... ) {
    return HELP_LEVEL_TICKET;
  } else {
    return HELP_LEVEL_FORUMS;
  }
}
```

southp replied:
Hmm, I'm not sure if that's possible for now, since `olark` is not in the redux store. Or is there a way that I overlooked?

jordwest replied:
Ah right, I forgot about that

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -418,12 +421,18 @@ const HelpContact = React.createClass( {
+		const olarkReadyOrTimedOut = olark.isOlarkReady && ! this.props.olarkTimedOut;
+		const ticketReadyOrError = isTicketSupportConfigurationReady || null !== ticketSupportRequestError;
+
+		const showPreloadForm = ! sitesInitialized || ! ticketReadyOrError || ! olarkReadyOrTimedOut ;
```
There are quite a lot of tricky boolean operations happening here that could probably be moved into selectors.

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -494,7 +503,20 @@ const HelpContact = React.createClass( {
+				{ shouldShowTicketRequestErrorNotice &&
+					<Notice
+						status='is-error'
+						text={ this.translate( 'An error occurred while requesting for email support settings. Please try again later.' ) }
```
> An error occurred while requesting ~~for~~ email support settings. Please try again later.

I'm not a big fan of the word 'error', it sounds very final. How about something like:

> We had trouble loading the support information for your account. Please check your internet connection and reload the page, or try again later.

southp replied:
I think your message is much well-put than mine, so I c&p it in f86599d3d5cdba9d7991c771770deab77f7a4475 :P

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -494,7 +503,20 @@ const HelpContact = React.createClass( {
+			<div>
+				{ shouldShowTicketRequestErrorNotice &&
+					<Notice
+						status='is-error'
```
What do you think about using a yellow warning notice instead?

southp replied:
Agree. Changed to `is-warning` by f86599d3d5cdba9d7991c771770deab77f7a4475

southp commented:
@jordwest I've restructured those boolean expressions in 771290bd9e4364026fe062dd06449d30bf87c209.
The general idea is to restructure the flow deciding the final contact form as a sequence of transformation: ( olark and email support eligibility flag ) => support variation slug => the properties of the contact form. Also, a bunch of other boolean expressions are also extracted in hope of helping the flow of `getView` look clearer.

Could you help me verify if that helps? cc @omarjackman on this since I've moved quite a bit around.

omarjackman commented on client/me/help/help-contact/index.jsx:
```diff
@@ -412,18 +421,119 @@ const HelpContact = React.createClass( {
+					showSiteField: moreThanOneSite,
+				};
+
+			case SUPPORT_FORUM:
```
This seems unnecessary

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -412,18 +421,118 @@ const HelpContact = React.createClass( {
+	shouldShowTicketRequestErrorNotice: function( variationSlug ) {
+		const { ticketSupportRequestError } = this.props;
+
+		return SUPPORT_HAPPYCHAT !== variationSlug && SUPPORT_LIVECHAT !== variationSlug && null !== ticketSupportRequestError;
```
For `null` checks it might be better to use `!=` instead of `!==`, since `null !== undefined` returns true. Or  just check for truthiness with `!!ticketSupportRequestError`.

southp replied:
Ah, good catch. This is something that I often forgot. Fixed by a003a1d504ee368a758e1dc8c6694834b9fcd156

jordwest commented on client/me/help/help-contact/index.jsx:
```diff
@@ -412,18 +421,118 @@ const HelpContact = React.createClass( {
+		const { ticketSupportConfigurationReady, ticketSupportRequestError } = this.props;
+
+		const olarkReadyOrTimedOut = olark.isOlarkReady && ! this.props.olarkTimedOut;
+		const ticketReadyOrError = ticketSupportConfigurationReady || null !== ticketSupportRequestError;
```
Null check as above

southp commented on client/me/help/help-contact/index.jsx:
```diff
@@ -412,18 +421,118 @@ const HelpContact = React.createClass( {
+
+		return SUPPORT_FORUM;
+	},
+
```
Agree. Fixed by 3ea8e4281389c4335aa3df77697a28c174fef9d4

jordwest commented:
LGTM 🚢
